tests/python/test_python_frontend.py

```diff
@@ -4458,6 +4457,140 @@ def fusion_func(fd: FusionDefinition):
+                is_cpu=False,
+            )
+            v = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            is_causal = fd.define_scalar(None, dtype=DataType.Bool)
+            scale = fd.define_scalar(None, dtype=DataType.Double)
+            attn, *intermediate_results = fd.ops.sdpfa_fwd(
+                q, k, v, is_causal=is_causal, scale=scale
+            )
+            fd.add_output(attn)
+
+        def fusion_default_causal(fd: FusionDefinition):
+            q = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            k = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            v = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            dropout_p = fd.define_scalar(None, dtype=DataType.Double)
+            scale = fd.define_scalar(None, dtype=DataType.Double)
+            attn, *intermediate_results = fd.ops.sdpfa_fwd(
+                q, k, v, dropout_p=dropout_p, scale=scale
+            )
+            fd.add_output(attn)
+
+        def fusion_default_scale(fd: FusionDefinition):
+            q = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            k = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            v = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            dropout_p = fd.define_scalar(None, dtype=DataType.Double)
+            is_causal = fd.define_scalar(None, dtype=DataType.Bool)
+            attn, *intermediate_results = fd.ops.sdpfa_fwd(
+                q, k, v, dropout_p=dropout_p, is_causal=is_causal
+            )
+            fd.add_output(attn)
+
+        N, H, L, S, E = 4, 8, 16, 16, 8
+        qkv = [
+            torch.randn((N, H, L, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+        ]
+
+        dropout_vals = [0.0, 0.2]
```

Comment:
:clap:

Reply:
nitpick on tests here.

I think the your default for None is just None?! so we can probably have it here as:
`dropout_vals = [0.0, 0.2, None]`. Same thing for `is_causal_vals` & `scale_vals`.

So we can just have `itertools.product(is_causal_vals, scale_vals, dropout_vals)`.

Sorry if my previous comment is misleading. We don't have to test pybind API behavior, we just need to cover all the code path in our part.

Reply:
Yes, I was attempting that, however scalars (defined through `fd.define_scalar`) does not accept `None` values. Is there any other way to represent `None` values  in fusions?

Reply:
Another way would be to pass these values to fusion definitions, however that triggers re-compilation failing checks in `exec_nvfuser`.

Reply:
ah... I see where it's coming from.

Would a partial work there?

```
def fusion_default(fd, has_dropout: bool, ...):

  dropout_p = None
  if has_dropout:
    dropout_p = fd.define_scalar(None, dtype=DataType.Double)
```

so you can have
```
  nvf_out, _ = self.exec_nvfuser(fusion_default_no_dropout = partial(fusion_default, False, ...), ...)
```

This way you can also simplify the script above.

Reply:
BTW, I'm not blocking your PR for this.

You mentioned that this test will be moved in follow up PR, so we can have the clean up then.

Reply:
Thanks, Jie. This was the way to go, the test looks much more concise now.